Mathlib/RingTheory/PolynomialLaw/Basic.lean

```diff
@@ -0,0 +1,265 @@
+finitely generated subalgebra is a quotient of a polynomial ring in the universe `u`, plus
+commutation of tensor products with direct limits, we will extend the functor to all `R`-algebras
+(to be done in a future PR).
+
+## Main definitions/lemmas
+
+* Instance : `Module R (M →ₚ[R] N)` shows that polynomial laws form a `R`-module.
+
+* `PolynomialLaw.ground f` is the map `M → N` corresponding to `PolynomialLaw.toFun' f R` under
+  the isomorphisms `R ⊗[R] M ≃ₗ[R] M`, and similarly for `N`.
+
+In further works, we construct the coefficients of a polynomial law and show the relation with
+polynomials (when the module `M` is free and finite).
+
+Reference : [Roby, Norbert. 1963. « Lois polynomes et lois formelles en théorie des modules ».
+Annales scientifiques de l’École Normale Supérieure 80 (3): 213‑348](Roby-1963)
+
+## Remark: Extension to commutative semirings
+
+In the literature, the theory is writen for commutative rings, but what is done here
+works as well for commutative semirings.
+
+-/
+
+universe u
+
+noncomputable section PolynomialLaw
+
+open scoped TensorProduct
+
+open MvPolynomial LinearMap TensorProduct AlgHom
+
+/-- A polynomial map `M →ₚ[R] N` between `R`-modules is a functorial family of maps
+   `S ⊗[R] M → S ⊗[R] N`, for all `R`-algebras `S`.
+
+For universe reasons, `S` has to be restricted to the same universe as `R`. -/
+@[ext]
+structure PolynomialLaw (R : Type u) [CommSemiring R]
+    (M : Type*) [AddCommMonoid M] [Module R M] (N : Type*) [AddCommMonoid N] [Module R N] where
+  /-- The functions `S ⊗[R] M → S ⊗[R] N` underlying a polynomial law -/
+  toFun' (S : Type u) [CommSemiring S] [Algebra R S] : S ⊗[R] M → S ⊗[R] N
+  /-- The compatibility relations between the functions underlying a polynomial law -/
+  isCompat' {S : Type u} [CommSemiring S] [Algebra R S]
+    {S' : Type u} [CommSemiring S'] [Algebra R S'] (φ : S →ₐ[R] S') :
+    φ.toLinearMap.rTensor N ∘ toFun' S = toFun' S' ∘ φ.toLinearMap.rTensor M
+
+/-- `M →ₚ[R] N` is the type of `R`-polynomial laws from `M` to `N`. -/
+notation:25 M " →ₚ[" R:25 "] " N:0 => PolynomialLaw R M N
+
+variable {R M N f} in
+theorem PolynomialLaw.isCompat_apply'
+    {R : Type u} [CommSemiring R]
+    {M : Type*} [AddCommMonoid M] [Module R M] {N : Type*} [AddCommMonoid N] [Module R N]
+    {f : M →ₚ[R] N}
+    {S : Type u} [CommSemiring S] [Algebra R S] {S' : Type u} [CommSemiring S'] [Algebra R S']
+    (φ : S →ₐ[R] S') (x : S ⊗[R] M) :
+    (φ.toLinearMap.rTensor N) ((f.toFun' S) x) = (f.toFun' S') (φ.toLinearMap.rTensor M x) := by
+  simpa only using _root_.congr_fun (f.isCompat' φ) x
+
+
+namespace PolynomialLaw
+
+
+section Module
+
+section CommSemiring
+
+variable {R : Type u} [CommSemiring R] {M : Type*} [AddCommMonoid M] [Module R M]
+  {N : Type*} [AddCommMonoid N] [Module R N] (r a b : R) (f g : M →ₚ[R] N)
+
+instance : Zero (M →ₚ[R] N) := ⟨{
+  toFun'    := fun _ => 0
+  isCompat' := fun _ => rfl }⟩
```

Comment:
By adding `attribute [local simp] isCompat_apply'` above, the proof of the naturality is found automatically:
```suggestion
instance : Zero (M →ₚ[R] N) := ⟨{ toFun' _ := 0 }⟩
```
(Same for `add` and `smul`.)